Sources/GRPC/CallHandlers/BidirectionalStreamingCallHandler.swift

```diff
@@ -33,62 +33,45 @@ public class BidirectionalStreamingCallHandler<
   public typealias EventObserver = (StreamEvent<RequestPayload>) -> Void
   public typealias EventObserverFactory = (Context) -> EventLoopFuture<EventObserver>
 
-  private var observerState: ClientStreamingHandlerObserverState<EventObserverFactory, EventObserver> {
-    willSet(newState) {
-      self.logger.debug("observerState changed from \(self.observerState) to \(newState)")
-    }
-  }
   private var callContext: Context?
+  private var eventObserver: EventLoopFuture<EventObserver>?
+  private let eventObserverFactory: (StreamingResponseCallContext<ResponsePayload>) -> EventLoopFuture<EventObserver>
 
   // We ask for a future of type `EventObserver` to allow the framework user to e.g. asynchronously authenticate a call.
   // If authentication fails, they can simply fail the observer future, which causes the call to be terminated.
   public init(
     callHandlerContext: CallHandlerContext,
     eventObserverFactory: @escaping (StreamingResponseCallContext<ResponsePayload>) -> EventLoopFuture<EventObserver>
   ) {
-    // Delay the creation of the event observer until `handlerAdded(context:)`, otherwise it is
```

Comment:
I don't think so: it's now no longer possible to make the observer until `processHead` because we don't pass it in the `CallHandlerContext`. The problem we had before was that the status promise could fail before the handler reached the pipeline; now it _must_ be in the pipeline when the observer is created.

Reply:
But the delaying part kind of still applies, doesn't it? That's what I was thinking about. I.e. just explain why we need a factory at all.

Reply:
Oh I see, makes sense to explain that then.